pkg/container-collection/options.go

```diff
@@ -252,23 +255,164 @@ func WithPubSub(funcs ...pubsub.FuncNotify) ContainerCollectionOption {
+		ownerRef = &ownerReferences[0]
+		if _, ok := expectedResKinds[ownerRef.Kind]; ok {
+			return ownerRef
+		}
+		return nil
+	}
+
+	for i, or := range ownerReferences {
+		if *or.Controller {
+			// There is at most one controller reference per resource
+			if _, ok := expectedResKinds[or.Kind]; ok {
+				return &or
+			}
+			return nil
+		}
+
+		// Keep track of the first expected reference in case it is be needed
+		if _, ok := expectedResKinds[or.Kind]; ownerRef == nil && ok {
+			ownerRef = &ownerReferences[i]
+		}
+	}
+
+	return ownerRef
+}
+
+func getKubeClientDynamic() (dynamic.Interface, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get Kubernetes config: %w", err)
+	}
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create dynamic Kubernetes client: %w", err)
+	}
+	return dynamicClient, nil
+}
+
+func getOwnerReference(dynamicClient dynamic.Interface,
+	resNamespace, resKind, resGroupVersion, resName string,
+) ([]metav1.OwnerReference, error) {
+	gv, err := schema.ParseGroupVersion(resGroupVersion)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse %s/%s groupVersion %s: %w",
+			resNamespace, resName, resGroupVersion, err)
+	}
+
+	params := schema.GroupVersionResource{
+		Group:    gv.Group,
+		Version:  gv.Version,
+		Resource: resKind,
+	}
+
+	res, err := dynamicClient.
+		Resource(params).
+		Namespace(resNamespace).
+		Get(context.TODO(), resName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch %s/%s %s/%s: %w",
+			resKind, resGroupVersion, resNamespace, resName, err)
+	}
+
+	return res.GetOwnerReferences(), nil
+}
+
+func ownerReferenceEnrichment(
+	containerDefinition *pb.ContainerDefinition,
+	ownerReferences []metav1.OwnerReference,
+) error {
+	dynamicClient, err := getKubeClientDynamic()
+	if err != nil {
+		return fmt.Errorf("failed get dynamic Kubernetes client: %w", err)
+	}
+
+	resGroupVersion := "v1"
+	resKind := "pods"
+	resName := containerDefinition.Podname
+	resNamespace := containerDefinition.Namespace
+
+	// Iterate until we reach the highest level of reference
+	for {
+		if len(ownerReferences) == 0 {
+			ownerReferences, err = getOwnerReference(dynamicClient,
+				resNamespace, resKind, resGroupVersion, resName)
+			if err != nil {
+				return fmt.Errorf("failed to get %s/%s/%s/%s owner reference: %w",
+					resNamespace, resKind, resGroupVersion, resName, err)
+			}
+
+			// No owner reference found
+			if len(ownerReferences) == 0 {
+				return nil
+			}
+		}
+
+		ownerRef := getExpectedOwnerReference(ownerReferences)
+		if ownerRef == nil {
+			// Any expected owner reference found
+			return nil
+		}
+
+		// Update container's owner reference
+		containerDefinition.OwnerReference = &pb.OwnerReference{
+			Apiversion: ownerRef.APIVersion,
+			Kind:       ownerRef.Kind,
+			Name:       ownerRef.Name,
+			Uid:        string(ownerRef.UID),
+		}
```

Comment:
I would like to but I lack some understanding of the code to be sure to do it without adding mistakes.
Indeed, we do not forcefully want to have `containerDefinition.OwnerReference` set to last `ownerRef` value.
If, for example, we break the loop because `ownerRef` is `nil` we will not be able to fill `containerDefinition.OwnerReference` using it.

For the moment, I will let it like this even though I am sure there is an optimization and I would like to avoid updating it at each iteration.

Reply:
I used a temporal variable to keep track of the highest reference.

Please let me know if it really made the code more readable.